guest/filter.go

```diff
@@ -24,57 +24,90 @@ import (
 	"sigs.k8s.io/kube-scheduler-wasm-extension/guest/api"
 	"sigs.k8s.io/kube-scheduler-wasm-extension/guest/internal/imports"
 	protoapi "sigs.k8s.io/kube-scheduler-wasm-extension/kubernetes/proto/api"
+	meta "sigs.k8s.io/kube-scheduler-wasm-extension/kubernetes/proto/meta"
 )
 
-// Filter should be assigned in `main` to a FilterFunc function.
+// FilterPlugin should be assigned in `main` to a FilterPlugin instance.
 //
 // For example:
 //
 //	func main() {
-//		guest.Filter = api.FilterFunc(nameEqualsPodSpec)
+//		guest.FilterPlugin = api.FilterFunc(nameEqualsPodSpec)
 //	}
-var Filter api.Filter
+var FilterPlugin api.FilterPlugin
 
 // filter is only exported to the host.
 //
 //go:export filter
 func filter() (code uint32) { //nolint
-	if Filter == nil {
+	if FilterPlugin == nil {
 		return
 	}
-	c, reason := Filter.Filter(filterArgs{})
-	if reason != "" {
+
+	// The parameters passed are lazy with regard to host functions. This means
+	// a no-op plugin should not have any unmarshal penalty.
+	// TODO: Make these fields and reset on pre-filter or similar.
+	s := FilterPlugin.Filter(&nodeInfo{}, &pod{})
+	if s == nil || s.Code == api.StatusCodeSuccess {
+		return uint32(api.StatusCodeSuccess)
+	}
+	if reason := s.Reason; reason != "" {
 		imports.StatusReason(reason)
 	}
-	return uint32(c)
+	return uint32(s.Code)
 }
 
-var _ api.FilterArgs = filterArgs{}
+var _ api.NodeInfo = (*nodeInfo)(nil)
 
-type filterArgs struct{}
+type nodeInfo struct {
+	n *protoapi.Node
+}
 
-func (filterArgs) NodeInfo() api.NodeInfo {
-	return nodeInfo{}
+func (n *nodeInfo) Node() *protoapi.Node {
+	return n.node()
 }
 
-func (filterArgs) Pod() *protoapi.Pod {
-	b := imports.Pod()
-	var msg protoapi.Pod
+func (n *nodeInfo) node() (node *protoapi.Node) {
+	if node = n.n; node != nil {
+		return
```

Comment:
Minor but I think using non-naked return makes this block easier to grok at the cost of just 5 letters

Reply:
agree!